pkg/docker/config/config.go

```diff
@@ -0,0 +1,274 @@
+	dockerCfgFileName = "config.json"
+	dockerLegacyCfg   = ".dockercfg"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the username and password in the auth.json file
+func SetAuthentication(ctx *types.SystemContext, registry, username, password string) error {
+	return modifyJSON(ctx, func(auths *dockerConfigFile) (bool, error) {
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, setAuthToCredHelper(ch, registry, username, password)
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+		return true, nil
+	})
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+// If an entry is not found empty strings are returned for the username and password
+func GetAuthentication(ctx *types.SystemContext, registry string) (string, string, error) {
+	if ctx != nil && ctx.DockerAuthConfig != nil {
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
+	}
+
+	dockerLegacyPath := filepath.Join(homedir.Get(), dockerLegacyCfg)
+	paths := [3]string{getPathToAuth(ctx), filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName), dockerLegacyPath}
+
+	for _, path := range paths {
+		legacyFormat := path == dockerLegacyPath
+		username, password, err := findAuthentication(registry, path, legacyFormat)
+		if err != nil {
+			return "", "", err
+		}
+		if username != "" && password != "" {
+			return username, password, nil
+		}
+	}
+	return "", "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(ctx *types.SystemContext, registry string) error {
+	return modifyJSON(ctx, func(auths *dockerConfigFile) (bool, error) {
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			return false, deleteAuthFromCredHelper(ch, registry)
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return false, ErrNotLoggedIn
+		}
+		return true, nil
+	})
+}
+
+// getPath gets the path of the auth.json file
+// The path can be overriden by the user if the overwrite-path flag is set
+// If the flag is not set and XDG_RUNTIME_DIR is ser, the auth.json file is saved in XDG_RUNTIME_DIR/containers
+// Otherwise, the auth.json file is stored in /run/user/UID/containers
+func getPathToAuth(ctx *types.SystemContext) string {
+	if ctx != nil {
+		if ctx.AuthFilePath != "" {
+			return ctx.AuthFilePath
+		}
+		if ctx.RootForImplicitAbsolutePaths != "" {
+			return filepath.Join(ctx.RootForImplicitAbsolutePaths, ctx.AuthFilePath)
```

Comment:
Actually, this isn’t the right place; here we know that `ctx.AuthFilePath == ""`.

Instead, apply `RootForImplicitAbsolutePaths` in the `runtimeDir == ""` fallback.

Reply:
If that is the case then we don't need to do `filepath.Join(ctx.RootForImplicitAbsolutePaths, ctx.AuthFilePath)` since `ctx.AuthFilePath == ""`?
Something like this then?
```
func getPathToAuth(ctx *types.SystemContext) string {
	if ctx != nil {
		if ctx.AuthFilePath != "" {
			return ctx.AuthFilePath
		}
	}
	runtimeDir := os.Getenv("XDG_RUNTIME_DIR")
	if runtimeDir == "" {
		if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
			return ctx.RootForImplicitAbsolutePaths
		}
		runtimeDir = filepath.Join(defaultPath, strconv.Itoa(os.Getuid()))
	}
	return filepath.Join(runtimeDir, authCfg, authCfgFileName)
}
```

Reply:
More like
```go
func getPathToAuth(ctx *types.SystemContext) string {
	if ctx != nil && if ctx.AuthFilePath != "" {
		return ctx.AuthFilePath
	}
	runtimeDir := os.Getenv("XDG_RUNTIME_DIR")
	if runtimeDir == "" {
		runtimeDir = filepath.Join(defaultPath, strconv.Itoa(os.Getuid()))
		if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
			runtimeDir = filepath.Join(ctx.RootForImplicitAbsolutePaths, runtimeDir)
		}
	}
	return filepath.Join(runtimeDir, authCfg, authCfgFileName)
}```

Reply:
okay, just curiosity, why filepath.Join ctx.RootForImplicitAbsolutePaths with runtimeDir?

Reply:
```go
// If not "", prefixed to any absolute paths used by default by the library (e.g. in /etc/).
```

Think of it as a c/image-scoped `chroot`: every time the library uses an absolute path which was not specified by the user, `RootForImplicitAbsolutePaths` is used instead of `/`.

This allows chroot-like uses with `/my/environment/with/prepared/etc` — where it’s clear enough that `/etc` should be in the chroot, but it is a tiny bit arguable whether `/run` should be in the chroot as well, or whether it should be considered an “API directory“ or whatever systemd likes to call it.

Another way to use this is for a library caller to essentially opt out of using to any files which haven’t been explicitly configured, by setting it to `/tmp/$uniquetemp/this/does/not/exist`, and by doing something similar to `$HOME`. In this case it seems clear enough that `/run` should be affected by the `RootForImplicitAbsolutePaths` option as well.


---

But, actually, the documentation of that option also says:
```go
// NOTE: If this is set, environment-variable overrides of paths are ignored (to keep the semantics simple: to create an /etc replacement, just set RootForImplicitAbsolutePaths .
// and there is no need to worry about the environment.)
```
so `RootForImplicitAbsolutePaths` should also override `XDG_RUNTIME_DIR`.

Reply:
oh okay. So I should check for `ctx.AuthFilePath`, then check for `ctx.RootForImplicitAbsolutePaths`, and then check for the environment variable `os.Getenv("XDG_RUNTIME_DIR")`?

Reply:
Yes.